From f9f34dcefc7ce96b618e6a8df4d8566081969019 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Mon, 20 Jun 2022 06:28:41 +0000
Subject: [PATCH] xen/arm: Implement MPAM domctls for L3 cache

This commit implements the domctls provided by xl-psr, namely the
XEN_DOMCTL_psr_alloc, to apply MPAM configs (currently only config to
control L3 cache) to a specific domain through "xl psr-cat-set" and
query current MPAM configs of a specific domain through
"xl psr-cat-show".

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-3763
Change-Id: Iccbd93c9e734a5d855f4f58cd0c4b2caac4efb5a
---
 xen/arch/arm/arm64/domctl.c | 14 ++++++++++++++
 xen/arch/arm/mpam.c         |  6 ++++--
 xen/include/asm-arm/mpam.h  |  1 +
 3 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/xen/arch/arm/arm64/domctl.c b/xen/arch/arm/arm64/domctl.c
index 04e12dcc7d..3a6f869ab0 100644
--- a/xen/arch/arm/arm64/domctl.c
+++ b/xen/arch/arm/arm64/domctl.c
@@ -8,6 +8,7 @@
 #include <xen/lib.h>
 #include <xen/errno.h>
 #include <xen/sched.h>
+#include <xen/guest_access.h>
 #include <xen/hypercall.h>
 #include <public/domctl.h>
 #include <asm/cpufeature.h>
@@ -61,6 +62,19 @@ long subarch_do_domctl(struct xen_domctl *domctl, struct domain *d,
     case XEN_DOMCTL_psr_alloc:
         if ( !cpu_has_mpam )
             return -EOPNOTSUPP;
+        switch ( domctl->u.psr_alloc.cmd )
+        {
+            case XEN_DOMCTL_PSR_SET_L3_CBM:
+                d->arch.mpam.cfg.l3_cpbm = domctl->u.psr_alloc.data;
+                return mpam_set_domain_resources(d);
+            case XEN_DOMCTL_PSR_GET_L3_CBM:
+                domctl->u.psr_alloc.data = d->arch.mpam.cfg.l3_cpbm;
+                if ( __copy_to_guest(u_domctl, domctl, 1) )
+                    return -EFAULT;
+                return 0;
+            default:
+                return -EOPNOTSUPP;
+        }
         return 0;
 #endif
 
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 1212ac0739..7e60067c8d 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -1440,7 +1440,7 @@ int mpam_get_hwinfo(enum mpam_class_types type, uint32_t data[],
 }
 
 /* TODO: Support other types of resource, e.g. SMMU. */
-static void mpam_set_domain_resources(struct domain *d)
+int mpam_set_domain_resources(struct domain *d)
 {
     uint32_t data[MPAM_INFO_ARRAY_SIZE] = { };
     unsigned int cpbm_lim = 0;
@@ -1450,7 +1450,7 @@ static void mpam_set_domain_resources(struct domain *d)
         if ( mpam_get_hwinfo(MPAM_CLASS_CACHE, data, ARRAY_SIZE(data)) )
         {
             printk(XENLOG_ERR "MPAM: L3 CPBM not supported for dom %pd\n", d);
-            return;
+            return -EINVAL;
         }
 
         cpbm_lim = (1 << data[MPAM_INFO_IDX_CPBM_LEN]) - 1;
@@ -1463,6 +1463,8 @@ static void mpam_set_domain_resources(struct domain *d)
 
     mpam_update_resources(&d->arch.mpam);
     smp_call_function(mpam_update_resources, &d->arch.mpam, 1);
+
+    return 0;
 }
 
 void __init mpam_presmp_setup(void)
diff --git a/xen/include/asm-arm/mpam.h b/xen/include/asm-arm/mpam.h
index b2796b6c72..507ca43c4e 100644
--- a/xen/include/asm-arm/mpam.h
+++ b/xen/include/asm-arm/mpam.h
@@ -438,6 +438,7 @@ extern void mpam_update_domain(struct domain *d);
 extern int mpam_discovery_cpu_online(unsigned int cpu);
 extern int mpam_get_hwinfo(enum mpam_class_types type, uint32_t data[],
                            unsigned int array_len);
+extern int mpam_set_domain_resources(struct domain *d);
 extern uint16_t mpam_get_system_max_partid(void);
 extern uint8_t mpam_get_system_max_pmg(void);
 
-- 
2.25.1

